src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -0,0 +1,147 @@
+package com.android.volley.toolbox;
+
+import android.annotation.SuppressLint;
+import com.android.volley.AsyncCache;
+import com.android.volley.VolleyLog;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.AsynchronousFileChannel;
+import java.nio.channels.CompletionHandler;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * AsyncCache implementation that uses Java NIO's AsynchronousFileChannel to perform asynchronous
+ * disk reads and writes. This should only be used by devices with an API level of 26 or above.
+ */
+@SuppressLint("NewApi")
+public class DiskBasedAsyncCache extends AsyncCache {
+
+    /** Map of the Key, CacheHeader pairs */
+    private final Map<String, CacheHeader> mEntries = new LinkedHashMap<>(16, .75f, true);
+
+    /** The supplier for the root directory to use for the cache. */
+    private final DiskBasedCacheUtility.FileSupplier mRootDirectorySupplier;
+
+    /** Total amount of space currently used by the cache in bytes. */
+    private long mTotalSize = 0;
+
+    /**
+     * Constructs an instance of the DiskBasedAsyncCache at the specified directory.
+     *
+     * @param rootDirectory The root directory of the cache.
+     */
+    public DiskBasedAsyncCache(final File rootDirectory) {
+        mRootDirectorySupplier =
+                new DiskBasedCacheUtility.FileSupplier() {
+                    @Override
+                    public File get() {
+                        return rootDirectory;
+                    }
+                };
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public synchronized void clear() {
+        // TODO (sphill99): Implement
+    }
+
+    /** Returns the cache entry with the specified key if it exists, null otherwise. */
+    @Override
+    public void get(String key, OnGetCompleteCallback callback) {
+        final OnGetCompleteCallback cb = callback;
+        final CacheHeader entry = mEntries.get(key);
+        // if the entry does not exist, return null.
+        if (entry == null) {
+            cb.onGetComplete(null);
+            return;
+        }
+        final File file = getFileForKey(key);
+        final int size = (int) file.length();
+        Path path = Paths.get(file.getPath());
+        AsynchronousFileChannel afc = null;
```

Comment:
I would move this inside the outer try, which should be a try-catch(IOException), and open a try-finally to close the afc on the next line:

```
try {
  final AsynchronousFileChannel afc = open()
  try {
    // I/O
  } finally {
    afc.close()
  }
} catch (IOException e) {
  // handle e
}
```

See https://github.com/google/volley/blob/master/src/main/java/com/android/volley/toolbox/DiskBasedCache.java#L148

Reply:
Since the class now requires a higher android api, I switched it to use a try-with-resources. I think this should allow it to just be a simple try-catch, and gets rid of these issues.